Scripts to subset full dataset from main into test

MattWellie opened the pull request:
`families_to_samples.py`; takes a project ID and a list of samples, prints out a dictionary of those families mapped to a list of their samples. Noted in the comments that if this is used on any regular basis, it should be migrated to sample-metadata where it can be a more efficient query

`extract_trio_vcf.py`; takes a project/dataset and the JSON string from the above script. Intended to run on main-bucket MT data to create a small subset (MT & VCF) we can use in test for prototyping. 

Note; this subsets the original MT in terms of samples, but not regions, so still quite large. Once this larger subset is in test, we'll be better able to minimise the data further without committing any specific region restriction into the main code branch. e.g. we can create a script on a scratch branch to extract only CHR22 for one sample, and then experiment with only that minimal dataset, but that isn't a logical script to commit to main.

see https://centrepopgen.slack.com/archives/C018KFBCR1C/p1641262742251800?thread_ts=1641251597.251300&cid=C018KFBCR1C for usage examples

illusional reviewed (commented):
Nice parameterization!!
I've added some basic code-style suggestions, but I'll leave the main script to @vladsaveliev as he will have a better idea on the the subsetting (but looks fine to me).

MattWellie commented on datasets/acute-care/extract_trio_vcf.py:
```diff
@@ -0,0 +1,60 @@
+    gcp_mt_full = os.path.join(gcp_main, "mt", f"{dataset}.mt")
+
+    # collect all unique sample IDs for a single filter on the MT
+    all_samples = set(chain.from_iterable(families_dict.values()))
```
I don't think this one works, as values() will be the list of lists, so the whole lot needs to be flattened rather than the lists being added to a set

MattWellie commented on datasets/acute-care/families_to_samples.py:
```diff
@@ -0,0 +1,195 @@
+    "--families",
+    "families",
+    type=click.STRING,
+    help="a comma-delimited string, containing all the family IDs to search for"
```
Nice improvement, changed and tested

MattWellie commented on datasets/acute-care/families_to_samples.py:
```diff
@@ -0,0 +1,195 @@
+    return resp
+
+
+def get_family_to_sample_map(df: pandas.DataFrame, families: str, external: bool, pid_to_id: dict, int_to_ext: dict
```
Nice improvement, changed and tested (with the click alteration)

MattWellie commented on datasets/acute-care/families_to_samples.py:
```diff
@@ -0,0 +1,195 @@
+
+    # currently the 2 types of endpoint require different headers, with the same auth token
+    table_header = CaseInsensitiveDict(data={"Accept": ACCEPT_ALL, "Authorization": f"Bearer {auth_token}"})
+    mapping_header = CaseInsensitiveDict(data={"Accept": ACCEPT_JSON, "Authorization": f"Bearer {auth_token}"})
```
I was checking the syntax for requests and I think their examples used CID (first time I've ever seen it). Truly useless in this situation, so it's removed

MattWellie commented on datasets/acute-care/families_to_samples.py:
```diff
@@ -0,0 +1,195 @@
+            if external:
+                dict_result[family_id].append(ext_sample_id)
+            else:
+                dict_result[family_id].append(int_sample_id)
```
Nice improvement, changed and tested

MattWellie commented on environment.yml:
```diff
@@ -5,4 +5,5 @@ dependencies:
-  - progressbar2
\ No newline at end of file
+  - progressbar2
+  - hail
```
Yup, that's a silly.

MattWellie commented:
Note: would like to add an assert that the provided participant IDs are all present in the aggregate prior to subsetting. Without this there is a risk that providing internal rather than external IDs (for example) would cause the creation of 'successful' completion and empty MT/VCFs

Goal - check for all participants requested being part of the aggregate:
- if not all are present, identify all absent as an error message, and exit
- if none are present, extract one example identifier and write into an exception to suggest to the user that they may have requested IDs in the wrong format

illusional commented:
Just a small clarification we're referring to sample IDs in the aggregate, and [this method](https://github.com/populationgenomics/fewgenomes/blob/b9694159fe79484feabb1480a84eb9d9f1175691/datasets/acute-care/families_to_samples.py#L171) is actually finding the _latest_ internal sample ID from the internal participant ID. Often we synthesize external participant IDs to be the same as the external sample ID, but this isn't always the case, and we do have cases where a single participant will have multiple samples (though that endpoint is maybe dangerously simplifying that).

MattWellie commented:
added an assertion in the script that all the samples we're looking for are present in the file, and explanatory error messages if this isn't the case. Also added unit tests to check functionality, which are executing fine locally.

@illusional I was a little worried about the latest/multiple sample IDs being a possible issue. I'm reasonably happy that the extraction input could be prepared locally & manually using Seqr instead of via that script, so you can safely ensure that the sample IDs of interest are being used (matching the pedigree samples displayed with the seqr loading pipeline input MT)

A more robust version of this process would take just a family ID and the PED file which was used in the pipeline when generating the aggregate for Seqr, and not bother with any DB queries...

illusional commented:
![image](https://user-images.githubusercontent.com/22381693/148465400-f50e3409-fbef-4df7-9487-c04fa828fd4b.png)

The method `get_external_participant_id_to_internal_sample_id` gets all active sample IDs for a list of external participant IDs. This might save you a few method calls anyway.

MattWellie commented:
We were considering flags to make an even more minimal subset, i.e. picking an individual chromosome. However, it appears that Hail/Spark _should_ handle operations behind the scenes so that only requested data is loaded for operations. In terms of persisting new files, the full data dump is more agile, and can be subsetted later on however we want once the data is in test.

Not yet sure about the compute/cost implications of 'loading' a full MT, but only querying a small portion... but theoretically it should be only the indices and requested data that's pulled from disk?

illusional commented on datasets/acute-care/extract_trio_vcf.py:
```diff
@@ -0,0 +1,171 @@
+    samples_in_mt = set(mat.s.collect())
+
+    # all good?
+    if not sample_names - samples_in_mt:
```
```suggestion
    if len(sample_names - samples_in_mt) == 0:
```

illusional commented on datasets/acute-care/extract_trio_vcf.py:
```diff
@@ -0,0 +1,171 @@
+    gcp_mt_full = os.path.join(gcp_main, "mt", f"{dataset}.mt")
+
+    # create the output paths as empty folders
+    os.makedirs(gcp_test_mt_outputs, exist_ok=True)
```
I don't think `os.makedirs` works with `gs://cpg-...` paths, it creates like a structure of nested directories.

```shell
python -c "import os; os.makedirs('gs://mfranklin/test/path')"

ls -lGh */**/
# gs:/:
# total 0
# drwxr-xr-x  3 michael.franklin  448689569    96B 10 Jan 13:13 mfranklin
# 
# gs:/mfranklin/:
# total 0
# drwxr-xr-x  3 michael.franklin  448689569    96B 10 Jan 13:13 test
# 
# gs:/mfranklin/test/:
# total 0
# drwxr-xr-x  2 michael.franklin  448689569    64B 10 Jan 13:13 path
# 
# gs:/mfranklin/test/path/:
```

MattWellie replied:
Agreed - I don't think it's going to behave as I expected so I'm pulling it out for now, and writing to the test root. Folder structure can be implemented later

illusional reviewed (approved):
Only the `os.makedirs` with a GCS path looks problematic, but otherwise looks great!
